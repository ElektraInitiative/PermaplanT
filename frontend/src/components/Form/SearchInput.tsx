import { ReactComponent as SearchIcon } from '@/icons/search.svg';
import { forwardRef, Ref } from 'react';

interface SearchInputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  /** input event handler */
  handleSearch: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const SearchInput = forwardRef(
  ({ handleSearch, ...inputProps }: SearchInputProps, ref: Ref<HTMLInputElement>) => {
    return (
      <div className="relative grow">
        <div className="pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3">
          <SearchIcon stroke="currentColor" />
        </div>
        <input
          onChange={handleSearch}
          ref={ref}
          {...inputProps}
          type="search"
          className="block h-11 w-full rounded-lg border border-neutral-500 bg-neutral-100 p-2.5 pl-10 text-sm placeholder-neutral-500 focus:border-primary-500 focus:outline-none disabled:cursor-not-allowed disabled:border-neutral-400 disabled:text-neutral-400 dark:border-neutral-400-dark dark:bg-neutral-50-dark dark:focus:border-primary-300 dark:disabled:border-neutral-400-dark dark:disabled:text-neutral-400-dark"
        />
      </div>
    );
  },
);

SearchInput.displayName = 'SearchInput';
export default SearchInput;
