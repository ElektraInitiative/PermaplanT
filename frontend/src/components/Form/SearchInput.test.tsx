import SearchInput, { SHORTCUT_SEARCH_INPUT_RESET, TEST_IDS } from './SearchInput';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ChangeEvent, ReactElement } from 'react';
import TestRenderer, { ReactTestRendererJSON } from 'react-test-renderer';

const mockHandleSearch = vi.fn();

describe('SearchInput', () => {
  it('should render correctly', () => {
    const jsonTree = TestRenderer.create(
      <SearchInput handleSearch={mockHandleSearch} />,
    ).toJSON() as ReactTestRendererJSON;

    expect(jsonTree).toMatchSnapshot();
  });

  it('should render the search icon', () => {
    renderComponent(<SearchInput handleSearch={mockHandleSearch} />);

    expect(screen.getByTestId(TEST_IDS.SEARCH_ICON)).toBeInTheDocument();
  });

  it('should render the search input field', () => {
    renderComponent(<SearchInput handleSearch={mockHandleSearch} />);

    expect(screen.getByTestId(TEST_IDS.SEARCH_INPUT)).toBeInTheDocument();
  });

  it('should only render the search reset icon when a search term is entered', async () => {
    const { user } = renderComponent(<SearchInput handleSearch={mockHandleSearch} />);

    expect(screen.queryByTestId(TEST_IDS.RESET_ICON)).not.toBeInTheDocument();

    await user.type(screen.getByTestId(TEST_IDS.SEARCH_INPUT), 'Tomato');

    expect(screen.getByTestId(TEST_IDS.RESET_ICON)).toBeInTheDocument();
  });

  it('should hide the search reset icon again when the search input field is manually cleared', async () => {
    const { user } = renderComponent(<SearchInput handleSearch={mockHandleSearch} />);

    await user.type(screen.getByTestId(TEST_IDS.SEARCH_INPUT), 'Tomato');

    expect(screen.getByTestId(TEST_IDS.RESET_ICON)).toBeInTheDocument();

    await user.clear(screen.getByTestId(TEST_IDS.SEARCH_INPUT));

    expect(screen.queryByTestId(TEST_IDS.RESET_ICON)).not.toBeInTheDocument();
  });

  it('should hide the search reset icon again when the search input field is cleared via click on reset icon', async () => {
    const { user } = renderComponent(<SearchInput handleSearch={mockHandleSearch} />);

    await user.type(screen.getByTestId(TEST_IDS.SEARCH_INPUT), 'Tomato');

    expect(screen.getByTestId(TEST_IDS.RESET_ICON)).toBeInTheDocument();

    await user.click(screen.getByTestId(TEST_IDS.RESET_ICON) as Element);

    expect(screen.queryByTestId(TEST_IDS.RESET_ICON)).not.toBeInTheDocument();
  });

  it('should clear the search input field when reset icon is clicked', async () => {
    const { user } = renderComponent(<SearchInput handleSearch={mockHandleSearch} />);
    const searchInputField = screen.getByTestId(TEST_IDS.SEARCH_INPUT);

    await user.type(searchInputField, 'Tomato');

    expect(searchInputField).toHaveValue('Tomato');

    await user.click(screen.getByTestId(TEST_IDS.RESET_ICON) as Element);

    expect(searchInputField).toHaveValue('');
  });

  it('should clear the search input field when shortcut is used', async () => {
    const { user } = renderComponent(<SearchInput handleSearch={mockHandleSearch} />);
    const searchInputField = screen.getByTestId(TEST_IDS.SEARCH_INPUT);

    await user.type(searchInputField, 'Tomato');

    expect(searchInputField).toHaveValue('Tomato');

    await user.keyboard(`{${SHORTCUT_SEARCH_INPUT_RESET}}`);

    expect(searchInputField).toHaveValue('');
  });

  it('should call passed search handler with current search term', async () => {
    let searchTerm = '';
    const handleSearch = (event: ChangeEvent<HTMLInputElement>) => {
      searchTerm = event.target.value;
    };
    const { user } = renderComponent(<SearchInput handleSearch={handleSearch} />);

    await user.type(screen.getByTestId(TEST_IDS.SEARCH_INPUT), 'Tomato');

    expect(searchTerm).toBe('Tomato');
  });
});

function renderComponent(component: ReactElement) {
  return {
    user: userEvent.setup(),
    ...render(component),
  };
}
