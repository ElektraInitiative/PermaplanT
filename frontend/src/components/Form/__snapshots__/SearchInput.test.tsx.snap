// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="relative grow"
>
  <div
    className="pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3"
  >
    <svg
      stroke="currentColor"
    />
  </div>
  <input
    className="block h-11 w-full rounded-lg border border-neutral-500 bg-neutral-100 p-2.5 pl-10 text-sm placeholder-neutral-500 focus:border-primary-500 focus:outline-none disabled:cursor-not-allowed disabled:border-neutral-400 disabled:text-neutral-400 dark:border-neutral-400-dark dark:bg-neutral-50-dark dark:focus:border-primary-300 dark:disabled:border-neutral-400-dark dark:disabled:text-neutral-400-dark"
    onChange={[MockFunction]}
    placeholder="Test"
    type="search"
  />
</div>
`;
